pkg/cmd/infra/router/template_test.go

```diff
@@ -0,0 +1,132 @@
+			description:        "When Action is `Set` it should not percent decode the incorrectly encoded value part Eg: percent is not followed by two hexadecimal digits",
+			inputValue:         "x-forwarded-client-cert:%25%7B%2BQ%7D%5Bssl_c_der%2Cbase64%5:Set",
+			expectError:        true,
+			expectedValue:      nil,
+			expectErrorMessage: "failed to decode percent encoding",
+		},
+		{
+			description:        "Invalid Action when header input spec has 2 parts",
+			inputValue:         "Accept-Encoding:Delete%2",
+			expectError:        true,
+			expectedValue:      nil,
+			expectErrorMessage: "invalid action Delete%2",
+		},
+		{
+			description:        "Invalid Action when header input spec has 3 parts",
+			inputValue:         "X-Frame-Options:DENY:Bad",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "invalid action Bad",
+		},
+		{
+			description:        "Should not allow specifying headers with no action or value",
+			inputValue:         "X-Frame-Options",
+			expectError:        true,
+			expectErrorMessage: "invalid HTTP header input specification: ",
+		},
+		{
+			description:        "Should not allow blank values for headers",
+			inputValue:         "",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "encoded header string not present",
+		},
+		{
+			description:        "Should fail with incorrect header name when HTTP header input specification has 3 parts",
+			inputValue:         "X-Frame-Options**!=:DENY:Set",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "invalid HTTP header name: X-Frame-Options**!",
+		},
+		{
+			description:        "Should fail with incorrect header name when HTTP header input specification has 2 parts",
+			inputValue:         "X-Frame-Options**!=:Delete",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "invalid HTTP header name: X-Frame-Options**!",
+		},
+		{
+			description:        "Should fail when header spec consists of more than 3 parts than Name:Value:Action",
+			inputValue:         "X-Frame-Options:DENY:Set:Invalid",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "invalid HTTP header input specification: X-Frame-Options:DENY:Set:Invalid",
+		},
+	}
+
+	for _, tc := range testCases {
+		switch gotValue, err := parseHeadersToBeSetOrDeleted(tc.inputValue); {
+		case !cmp.Equal(gotValue, tc.expectedValue):
+			t.Errorf(" expected %s, got %s", tc.expectedValue, gotValue)
+		case err == nil && tc.expectError:
+			t.Errorf("%s: expected error, got nil", tc.description)
+		case err != nil && !tc.expectError:
+			t.Errorf("%s: expected success, got error: %v", tc.description, err)
+		case err != nil && len(tc.expectErrorMessage) != 0:
+			if !strings.Contains(fmt.Sprint(err), tc.expectErrorMessage) {
+				t.Errorf("the error message %s returned does not match expected return message %s", err, tc.expectErrorMessage)
+			}
+		}
```

Comment:
Do this like https://github.com/openshift/library-go/pull/1449/files#diff-51cbcb9e560e512d7dba0cd93dab386055814a27e2c7903ad93bce92c58fd18bR1680, something more like:
```suggestion
                var actualErrorMessage string
                if err != nil {
                    actualErrorMessage = err.Error()
                }
		switch {
			case tc.expectedErrorMessage == "" && actualErrorMessage != "":
				t.Fatalf("unexpected error: %v", actualErrorMessage)
			case tc.expectedErrorMessage != "" && actualErrorMessage == "":
				t.Fatalf("got nil, expected %v", tc.expectedErrorMessage)
			case !strings.Contains(actualErroMessage, tc.expectedErrorMessage):
				t.Fatalf("unexpected error: %v, expected: %v", actualErrorMessage, tc.expectedErrorMessage)
			}
```

Reply:
yes

Reply:
done